### DISCUSSION TITLE
reducev performance

### DISCUSSION THREAD

**Discussion by kleisauke** [libvips/libvips#2757] [Help wanted]
The extra `vips_sequential` in `reducev`, introduced in commit https://github.com/libvips/libvips/commit/3ca206815cfbe69d3387aa9f3bd360cadd04001a, could slow down multi-threaded resize operations. This may cause `--vips-concurrency 1` or `VIPS_CONCURRENCY=1` to perform better than the same operation executed with full concurrency. See for example this gist:
https://gist.github.com/kleisauke/7d6f194cde52737ce8a494ee0a5e55e7

On that gist, removing this extra `vips_sequential` makes the C-path of `reducev` ~67% faster, and the AVX2/SSE4.1 paths (see the [`simd`](https://github.com/kleisauke/libvips/tree/simd) branch) varying between ~40% and ~52% faster. The multi-threaded timings are all inline after that, with at most 16ms difference.

This discussion was opened to find possible alternatives for resolving the "out of order read" errors under heavy load, which were fixed with that commit.

(This was [initially discussed on Gitter](https://gitter.im/libvips/devchat?at=6103eb286ccf813819e6ce31), but I thought it would be better to track it here)

**Comment by jcupitt** [comment]
I agree, that was a really expensive fix, there must be something better.

I think this is where that extra seq came from: https://github.com/libvips/libvips/issues/617 ie. it was part of the revised seq mechanism.

It'd be good to get a reproducer for this problem so we can test possible improvements. I've tried commenting out the extra seq and trying to provoke an out of order read, but not managed it yet.

There's this issue as well from a month later: https://github.com/libvips/libvips/issues/639 that one finds and fixes a couple of small cache size issues, so (perhaps??) it makes the addition of the seq unnecessary now?

I've tried using `VIPS_STALL`, ie.:

```
$ VIPS_STALL=1 vips reducev k2.jpg x.jpg 1.5
$ VIPS_STALL=1 vips reducev st-francis.jpg x.jpg 1.5
$ VIPS_STALL=1 vips reducev st-francis.jpg x.jpg 16
```

But I've not been able to provoke an out of order read.

**Reply by kleisauke** [reply]
After removing the extra seq, I could reproduce this quite easily with:
```
$ vips black x.jpg 539 960
$ VIPS_STALL=1 vips reducev x.jpg x.v 2
vips_task_work_unit: stall done, releasing y = 0 ...

(vips:5311): VIPS-WARNING **: 12:35:50.756: error in tile 0 x 248

(vips:5311): VIPS-WARNING **: 12:35:50.756: error in tile 0 x 256

(vips:5311): VIPS-WARNING **: 12:35:50.756: error in tile 0 x 264

(vips:5311): VIPS-WARNING **: 12:35:50.756: error in tile 0 x 272

(vips:5311): VIPS-WARNING **: 12:35:50.757: error in tile 0 x 280

(vips:5311): VIPS-WARNING **: 12:35:50.757: error in tile 0 x 288
vips_task_work_unit: stall done, releasing y = 0 ...
vips_task_work_unit: stall done, releasing y = 0 ...
VipsJpeg: out of order read at line 776
```

or:
```
$ vips black x.jpg 10000 10000
$ VIPS_STALL=1 vips reducev x.jpg x.v 2
vips_task_work_unit: stall done, releasing y = 0 ...

(vips:5340): VIPS-WARNING **: 12:36:06.491: error in tile 0 x 248

(vips:5340): VIPS-WARNING **: 12:36:06.492: error in tile 0 x 256

(vips:5340): VIPS-WARNING **: 12:36:06.492: error in tile 0 x 264

(vips:5340): VIPS-WARNING **: 12:36:06.492: error in tile 0 x 272

(vips:5340): VIPS-WARNING **: 12:36:06.492: error in tile 0 x 280

(vips:5340): VIPS-WARNING **: 12:36:06.492: error in tile 0 x 288
vips_task_work_unit: stall done, releasing y = 0 ...
vips_task_work_unit: stall done, releasing y = 0 ...
VipsJpeg: out of order read at line 776
```

**Reply by jcupitt** [reply]
Ah that's a shame. But good to have a reproducer, nice!

**Reply by jcupitt** [reply]
That doesn't reproduce on this PC :( I'll try a laptop.

**Reply by kleisauke** [reply]
Ah, I could not reproduce this with `--vips-concurrency 32` (IIRC, that's the same number of threads of your threadripper workstation). However, I was able to reproduce this using: `--vips-concurrency 6` (which is the same number of threads of my i5-8600K workstation).

**Reply by kleisauke** [reply]
Here's another benchmark: https://gist.github.com/kleisauke/8e27833e3c0ac0b4f35200203cc36bf4 (on Windows with a 5000x5000 image).

It's curious to see that this performance degradation affects the `simd` branch more than the `master` branch. Perhaps the `THINSTRIP` -> `FATSTRIP` change could cause that (commit https://github.com/kleisauke/libvips/commit/ef2f42f2e12b2a7215887a9317fe4873c132432f)?

**Reply by kleisauke** [reply]
I've updated the above benchmark gist for libvips 8.13.0 and latest `simd` branch revision. Here are some handy commits to test with:
https://github.com/kleisauke/libvips/commit/735ac0a559f4f4d567f513180b741083552fb216 (based on the `master` branch)
https://github.com/kleisauke/libvips/commit/6733cd4e209142d77654710746cda892b61ad59b (based on the `simd` branch)

**Comment by jcupitt** [comment]
There's an issue and sample code here with a possible improvement https://github.com/libvips/libvips/issues/3754

**Reply by jcupitt** [reply]
To save you a click, the idea is that we should have some thread-private storage containing an allocation number -- a small integer assigned in order from 0 by the threadpool allocation function (this can go in the base class) and order by that. sequential should stall threads which are out of order, and only let them though by strict allocation number.

This means out of order reads should be impossible, except in the cases where there's very large non-locality.
